File: app/src/main/java/com/udindev/sade/cobacoba/FavoriteRepository.java
package com.udindev.sade.cobacoba;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.ArrayList;
import java.util.List;

public class FavoriteRepository {
    private final String TAG = getClass().getSimpleName();
    private final FirebaseFirestore database = FirebaseFirestore.getInstance();

    private final MutableLiveData<Favorite> resultData = new MutableLiveData<>();
    public LiveData<Favorite> getData(){
        return resultData;
    }

    public void query(String userId){
        database.collection("favorite").document(userId)
                .get()
                .addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                        if (task.isSuccessful()){
                            Favorite favorite = task.getResult().toObject(Favorite.class);

                            // Pengguna yang belum pernah tambah favorit
                            if (favorite == null) {
                                favorite = new Favorite(new ArrayList<>());
                                insert(userId, favorite);  // Sekalian bikin dokumennya
                            }

                            resultData.postValue(favorite);
                        } else Log.w(TAG, "Error querying document", task.getException());
                    }
                });
    }

    private void insert(String userId, Favorite favorite){
        database.collection("favorite").document(userId)
                .set(favorite)
                .addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        if (task.isSuccessful()) Log.d(TAG, "Document was added");
                        else Log.w(TAG, "Error adding document", task.getException());
                    }
                });
    }

    public void add(String userId, String productId){
        database.collection("favorite").document(userId)
                .update("listProductId", FieldValue.arrayUnion(productId))
                .addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        if (task.isSuccessful()) Log.d(TAG, "Document was updated");
                        else Log.w(TAG, "Error updating document", task.getException());
                    }
                });
    }

    public void remove(String userId, String productId){
        database.collection("favorite").document(userId)
                .update("listProductId", FieldValue.arrayRemove(productId))
                .addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        if (task.isSuccessful()) Log.d(TAG, "Document was updated");
                        else Log.w(TAG, "Error updating document", task.getException());
                    }
                });
    }
}

File: app/src/main/java/com/udindev/sade/repository/FavoriteRepository.kt
package com.udindev.sade.repository

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.firebase.firestore.FirebaseFirestore
import com.udindev.sade.model.Favorite

class FavoriteRepository {

    private val db = FirebaseFirestore.getInstance()

    private var resultDataEmail : MutableLiveData<List<Favorite>> = MutableLiveData()
    private var resultData: MutableLiveData<List<Favorite>> = MutableLiveData()
    fun getResults(): LiveData<List<Favorite>> = resultData

    fun getResultsEmail(): LiveData<List<Favorite>> = resultDataEmail

    fun getData() {
        val produkData: MutableList<Favorite> = ArrayList()
        val db = FirebaseFirestore.getInstance()
        val savedProdukList = ArrayList<Favorite>()
        db.collection("favorite")
                .get()
                .addOnSuccessListener { result ->
                    for (document in result) {
                        val pp = document.toObject(Favorite::class.java)
                        pp.id = document.id
                        savedProdukList.add(pp)
                        produkData.add(pp)
                        Log.d(TAG, "getData size : ${savedProdukList.size}")
                        Log.d(TAG, "getData: $pp ")
                    }
                    resultData.value = produkData
                    Log.d(TAG, "readProduk size final getData : ${savedProdukList.size}")
                }
                .addOnFailureListener { exception ->
                    Log.e(TAG, "Error getting documents.", exception)
                }
    }



    fun addFavorite(favorite: Favorite) {
        val item = hashMapProduk(favorite)
        db.collection("favorite")
                .add(item)
                .addOnSuccessListener { documentReference ->
                    Log.d(TAG, "DocumentSnapshot added with ID: ${documentReference.id}")
                }
                .addOnFailureListener { e ->
                    Log.e(TAG, "Error adding document", e)
                }
    }

    fun deleteFavorite(favorite: Favorite) {
        val idProduk: String? = favorite.id
        hashMapProduk(favorite)
        if (idProduk != null) {
            db.collection("favorite").document(idProduk)
                    .delete()
                    .addOnSuccessListener {
                        Log.d(TAG, "DocumentSnapshot successfully deleted!")
                    }
                    .addOnFailureListener { e ->
                        Log.w(TAG, "Error deleting document", e)
                    }
        }
    }

    fun getDataEmail(email: String) {
        val produkData: MutableList<Favorite> = ArrayList()
        val db = FirebaseFirestore.getInstance()
        val savedProdukList = ArrayList<Favorite>()
        db.collection("favorite")
                .get()
                .addOnSuccessListener { result ->
                    for (document in result) {
                        val namaDocument = document.data["email"] as String
                        if (namaDocument == email ) {
                            val pp = document.toObject(Favorite::class.java)
                            pp.id = document.id
                            savedProdukList.add(pp)
                            produkData.add(pp)
                            Log.d(TAG, "getDataEmail Favorite size : ${savedProdukList.size}")
                            Log.d(TAG, "getDataEmail Favorite: $pp ")
                        }
                    }
                    resultDataEmail.value = produkData
                    Log.d(TAG, "readProduk size final getDataEmail Favorite : ${savedProdukList.size}")
                }
                .addOnFailureListener { exception ->
                    Log.e(TAG, "Error getting documents.", exception)
                }
    }


    private fun hashMapProduk(favorite: Favorite): HashMap<String, Any?> {
        return hashMapOf(
                "idDocument" to favorite.idDocument,
                "email" to favorite.email
        )
    }

    companion object {
        private const val TAG = "FavoriteRepository"
    }
}